AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122434
* #122981
* #123646
* __->__ #122433

bdhirsh committed:
AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases

[ghstack-poisoned]

bdhirsh commented:
cc @yf225 this should fix the "input mutation not showing up in the graph" stuff we were seeing with FSDP

cc @oulgen in case you're interested in reviewing (this touches some of the "hide triton mutations from autograd" stuff we did last half)

ezyang commented:
I'm happy to stamp this but I admit I have completely forgotten (or perhaps I never knew?) how the functional tensor "did we mutate under autograd" logic worked at all, so while the change seems plausible I don't have any intuition for what it's supposed to do.

xmfan commented on aten/src/ATen/FunctionalStorageImpl.h:
```diff
@@ -103,6 +103,31 @@ struct TORCH_API FunctionalStorageImpl : public c10::StorageImpl {
+  bool are_all_mutations_hidden_from_autograd() const {
+    // mutations under no_grad / inference_mode are technically not hidden from
+    // autograd - they change the version counter
+    return mutation_counter_ <= mutation_counter_hidden_from_autograd_;
```
is mutation_counter incremented each time we increment mutation_counter_hidden_from_autograd_?

bdhirsh replied:
The only place this is used to today is for triton kernels (which hide all mutations from autograd).

There's a HOP for triton kernels, that calls this API [here](https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/triton_kernel_wrap.py#L808). And it's carefully written to bump both of those mutation counters by the same amount (which is an invariant we need to make sure we maintain in any future places where we need to call this API).

ezyang replied:
Can we make illegal states unrepresentable? Instead of `mark_mutation` have `mark_mutation_sees_autograd`. Now they're all disjoint and you test if certain fields are zero or not.

ezyang replied:
I guess part of the reason is that these were preexisting lol

bdhirsh replied:
yeah, part of the problem is that the triton HOP (linked above) has to call `replace_()` in its functionalization rule, which will implicitly bump the `mutation_counter_`.

One option is that I add a kwarg to `replace_(hidden_from_autograd=False)`, so `replace_()` is the one API that everyone goes through to bump mutation counters.

bdhirsh commented:
> how the functional tensor "did we mutate under autograd" logic worked at all, so while the change seems plausible I don't have any intuition for what it's supposed to do.

This improved incrementally over the last couple months, so let me try to give more background.

The logic for "should we include an input mutation inside the graph vs. leave it in an opaque epilogue" has historically been pretty complicated, so it was recently simplified to be "all input mutations can stay in the graph, unless the input tensor `requires_grad` **and** the mutation affects the autograd metadata of of the input in a non-trivial way (mutates the .grad_fn).

This means that input mutations on tensor that require grad can technically stay in the graph in the following two cases:

(1) every mutation that happened to the tensor was under no_grad or inference_mode. In this case, the only autograd metadata that changes is the version counter. So if we can detect this properly, and we keep the mutation in the graph, then we just need to probably bump the VC in our `autograd.Function` (with `mark_dirty`)

(2) every mutation that happened to the tensor was due to a triton kernel. Triton kernel mutations are 100% hidden from autograd, including the version counter not getting bumped, so if can detect that **every** mutations that happened to an input was from a triton kernel, then we basically have to do nothing at runtime.

The way we currently detect (1) in functionalization is: every mutation kernel in functionalization eventually calls `FunctionalTensorWrapper::replace_()`. `replace_()` peaks at the current autograd state to see if no_grad or inference_mode are active, so it can record whether or not the current mutation is running under no_grad/infrence_mode.

This was not quite right though: because `replace_()` is also called when we lazily regenerate "updated aliases". For example:
```
y = x.view(-1)
with torch.no_grad():
    y.mul_(2)
return x + 1
```
When functionalization sees the `x + 1`, it runs the logic to regenerate the "updated x", given "updated y". This also calls `replace_()`, but at this point `no_grad()` is no longer active, so we (incorrectly) mark `x` as having experienced a mutation outside of the no_grad region.

I tried to fix this in the PR with two changes:

(a) I took the counters that I use to track how many mutations happened under no_grad or not and moved them to the FunctionalTensor storage. My reasoning was that "whether or not a tensor was mutated and in what context" is a property that is shared across a tensor and all of its aliases.

Thinking about it more, maybe that is... too conservative? If you mutate a tensor under no_grad, and mutate its alias outside of no_grad, we will pessimistically assume that both tensor mutations cannot be kept in the graph. I did find this model a bit simpler to reason about though, and this case should be pretty rare

(b) I added a flag to `replace_()` so we know when it is getting called during the view-regeneration logic, so I know to only bump counters when we are not doing view regeneration.

@ezyang let me know if any of that is unclear / needs more detail

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

ezyang commented on test/functorch/test_aotdispatch.py:
```diff
@@ -731,6 +731,23 @@ def f(a):
+            return a + 1
+        inp = [torch.ones(4, requires_grad=True)]
+        fw_graph = self.verify_aot_autograd(f, inp, test_mutation=True, only_keep_inference_mutations=True)
+        self.assertExpectedInline(fw_graph.code.strip(), """\
```
In tests like this it's helpful to mention what exactly we're looking for in the graph. In this case, I believe it's that there should be a copy_ in the graph

bdhirsh replied:
Definitely agreed, I'll call it out (the important bit is indeed that we get a copy_() inside the graph, and not in the opaque epilogue)

ezyang commented on aten/src/ATen/FunctionalStorageImpl.h:
```diff
@@ -121,6 +146,10 @@ struct TORCH_API FunctionalStorageImpl : public c10::StorageImpl {
+
+  uint64_t mutation_counter_during_no_grad_or_inference_mode_ = 0;
+  uint64_t mutation_counter_ = 0;
+  uint64_t mutation_counter_hidden_from_autograd_ = 0;
```
these want docs, basically the explanation you posted in the comment

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -231,10 +231,16 @@ void FunctionalTensorWrapper::replace_(const Tensor& other) {
+  // if a mutation happens to a view under a no_grad,
+  // we won't call replace_() on the other alias until the alias is later used, which
+  // might not be until after the no_grad region is exited.
+  // Therefore, replace_() is not unconditionally safe to check the current no_grad state.
```
"However, if this is a lazy regeneration, then it is guaranteed that we have already done the mutation for the storage alias (when we originally performed the mutation), so no counter update may be needed."

Also, we can overreport this way, right? E.g., if I do a mutation on alias and I immediately replace_ the view and base, that will be +2 on the counters. Seems harmless for what you're using it for, but seems worth noting.

bdhirsh replied:
> "However, if this is a lazy regeneration, then it is guaranteed that we have already done the mutation for the storage alias (when we originally performed the mutation), so no counter update may be needed."

yep this is clearer, will add

> f I do a mutation on alias and I immediately replace_ the view and base, that will be +2 on the counters

Hmm I'm not sure I follow. If I mutate an FunctionalTensorWrapper that is a base of some other FunctionalTensorWrapper, then the current setup looks like:

```
base = torch.ones(2)
view = base.view(-1)
view.mul_(2) # updates "view"

return base + base # updates "base" view lazy regenerate
```


<img width="829" alt="image" src="https://github.com/pytorch/pytorch/assets/16311747/68b2486e-e30a-4c5f-9622-66122e773c6d">

Where the following would run:

(1) `view.mul_(2)`: this calls `replace_()` exactly once, on the "view" tensor.

(2) `return base + base`: this calls `base->sync_()`, which runs the lazy regenerate logic to regenerate `base` off of the underlying "updated base tensor" in the `FunctionalStorageImpl`. In this case, there are zero views in the view chain between the underlying storage and `base`.

ezyang replied:
Oh, I thought we immediately ran off and updated base sync (because the view has access to base), then I guess this does not happen.

ezyang reviewed (approved):
This doesn't seem harmful, but there's some improvements I think are worth investigating.

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

ezyang commented on aten/src/ATen/FunctionalStorageImpl.h:
```diff
@@ -103,6 +103,31 @@ struct TORCH_API FunctionalStorageImpl : public c10::StorageImpl {
+        mutation_counter_during_no_grad_or_inference_mode_ +
+        mutation_counter_hidden_from_autograd_;
+    // The <= is because both counters will technically be incremented, if we
+    // perform e.g. a triton kernel mutation under no_grad
```
Blegh. I guess you could have hidden_from_autograd_and_under_no_grad. What a pain.

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: fix detection for mutations under no_grad when there are outstanding aliases"

Fixes https://github.com/pytorch/pytorch/issues/122436.

The problem was that even though we were detecting when mutations happen under no_grad or not, we were recording these mutations when they happened to the FunctionalTensor - we should really just be recording them on the underlying storage.

In particular, what would happen is that we would mutate an alias under no_grad (marking the mutation as under no_grad properly), but if we use the base tensor outside of the no_grad region, we would lazily regenerate the base at this point, propagate the mutation to the base, and at that point mark that the base witnessed a mutation (outside of the no_grad region)




[ghstack-poisoned]
